File: Dynamic Programming/*New 21 Game/solution.py
class Solution:
    def new21Game(self, n: int, k: int, w: int) -> float:
        if k == 0 or k + w < n:
            return 1 
        dp = [0 for _ in range(n+1)]
        dp[0] = 1
        # dp[i] = dp[i-1] * 1/n + dp[i-2] * 1/n + ... + dp[i-w] * 1/n
        # => dp[i] = 1/n * sum(dp[i] for i in range(1, w+1))
        s = 1
        for i in range(1, n+1):
            dp[i] = s * 1/w
            if i < k:
                s += dp[i]
            if i >= w:
                s -= dp[i-w]
        return sum(dp[i] for i in range(k, n+1))

File: Dynamic Programming/*New 21 Game/solution.rs
impl Solution {
    pub fn new21_game(n: i32, k: i32, w: i32) -> f64 {
        if k == 0 || k + w < n {
            return 1.0;
        }
        let mut dp = vec![0 as f64; (n+1) as usize];
        dp[0] = 1.0;
        let mut s = 1.0;
        for i in 1..n+1 {
            dp[i as usize] = s * 1.0 / w as f64;
            if i < k {
                s += dp[i as usize];
            }
            if i >= w {
                s -= dp[(i-w) as usize];
            }
        }
        return dp[k as usize..(n+1) as usize].iter().sum();
    }
}
